### DISCUSSION TITLE
Are there any methods to extract ONLY the visible content after the filter is applied?

### DISCUSSION THREAD

**Discussion by amalgame21** [andmarti1424/sc-im#673] [Q&A]
I use sc-im to deal with csv files now, I found that it is very useful.
But when I apply a filter, say :addfilter "#A=1"
Those row match this filter (A not equal 1) will be filtered out, remaining some visible cells.
But I cannot find any method to extract ONLY those visible cells, like copy and paste, or use ":e csv" to export to other file
The hidden cells or rows are copy and paste, or export together with the visible one if visual mode is used.
I cannot find any means to separate them to the other cells, different sheets or different files.
Thanks!

**Comment by 097115** [comment]
Not sure if I understand you correctly, so correct me if I'm wrong, but these actions seem to produce filtered mkd and csv files for me:

  - `sc-im test.csv` (let's say, this [test.csv](http://ix.io/3NPZ))
  - `:addfilter "#A<25"`
  - `:filteron A0:C100`
  - `:e mkd` # results in `File "test.mkd" written`
  - `:e csv test_filtered.csv` # results in `File "test_filtered.csv" written`

**Reply by amalgame21** [reply]
> Not sure if I understand you correctly, so correct me if I'm wrong, but these actions seem to produce filtered mkd and csv files for me:
> 
> * `sc-im test.csv` (let's say, this [test.csv](http://ix.io/3NPZ))
> * `:addfilter "#A<25"`
> * `:filteron A0:C100`
> * `:e mkd` # results in `File "test.mkd" written`
> * `:e csv test_filtered.csv` # results in `File "test_filtered.csv" written`

oh maybe I describe it not extensive enough.


To reproduce what I experienced
Also use the [test.csv](http://ix.io/3NPZ) as an example
lets use the filter `:addfilter "@eqs(@substr($B,1,1),'B')"` instead to only filter words in column B start with letter "B"
`:filteron A0:C100`
only row 5, 13, 14, 15, 17, 59 is visible
`:e mkd` # results in `File "test.mkd" written`
`:e csv test_filtered.csv` # results in `File "test_filtered.csv" written`

The content of the test.mkd and test_filtered.csv contain the complete 60 rows of the original test.csv, not only 5,13,14,15,17,59 rows
Even if I use visual mode to select those 5, 13, 14, 15, 17, 59 rows after applied the filter, and copy and paste to other cells, it copy the whole 5-59 rows.
Using visual mode selecting 5, 13, ,14, 15, 17, 59 row and export to other format like mkd or csv, the result also contain the 5-59 rows.

**Reply by 097115** [reply]
Aha! Indeed, you are right, @amalgame21. Seems like the filtering system is... complicated :) And unless I'm missing something, it looks like a bug to me :)

A (partly) related report was submitted recently: https://github.com/andmarti1424/sc-im/issues/668. You may want to add the details in the comments or may be to open another ticket, since `sc-im` clearly looses a filter in certain (most?) cases.

**Reply by amalgame21** [reply]
Thanks for your advice! I opened another ticket #674, thanks!

**Reply by 097115** [reply]
Thanks, subscribed to your issue :)

Also, you sure know it, but just on case: we can easily filter csvs with `awk`, for example:

    awk -v FS=',' '{if ($2 ~ "^B") {print $0}}' test.csv > test_filtered.csv

It's a pity, of course, that we can't pipe the result to `sc-im` directly but it's still a workaround until the the filters get fixed :) 

(The original [`sc`](https://github.com/n-t-roff/sc/) is [said](https://github.com/n-t-roff/sc/blob/ba00dc7877ee2826bf0d75e7563f2aa1bb5352eb/sc.doc#L51) to be able to read from stdin but I, too, failed to make this work.)

**Reply by amalgame21** [reply]
Oh I didn't know it!
awk syntax it still too complicated for me, your example is a good reference for me to start with those filter processing.
thanks for your advice!

I read the `:help` menu of `sc-im` and it said it can pipe the result to sc-im with the `--nocurses` flag
but still no success, just displaying a lot of `syntax error`
and I still do not know how to use non-curses mode

**Reply by 097115** [reply]
`--nocurses` disables the [TUI](https://en.wikipedia.org/wiki/Text-based_user_interface), so we could use `sc-im` as a command-line app, which indeed would allow some sort of piping.

Let's say, we have the following sheet `test.sc`

    set external_functions
    let A0 = 10
    let B0 = @ston(@ext("echo", A0))
    recalc
    getnum B0

We can open it as usual, `sc-im test.sc`. Or, we can just print the requested result (`getnum A1`) to stdin:

    cat test.sc | sc-im --nocurses

optionally also adding the `--quit_afterload` parameter, or, for instance, if we'd just like to export our table, also adding `--quiet` to suppress any system information:

    echo 'export "mkd" "/dev/stdout"' | sc im --nocurses --quiet --quit_afterload test.sc

As for `awk`, the basic intro is not very big ([for example](https://www.linuxcommands.site/linux-text-processing-commands/linux-awk-command/awk-syntax/)), so it's really worth diving in. Besides, for **exploring** data, [visidata](https://github.com/saulpw/visidata) is also quite good (there's a [nice tutorial](https://jsvine.github.io/intro-to-visidata/), e.g., pages on filtering: [1](https://jsvine.github.io/intro-to-visidata/basics/sorting-and-filtering/#filtering-selected-rows-with), [2](https://jsvine.github.io/intro-to-visidata/practical/high-flying-birds/#select-only-rows-where-height-is-greater-than-zero)). And then, there's also [R](https://en.wikipedia.org/wiki/R_(programming_language)) :)
